### ISSUE TITLE
Don't store passwords in clear text

### ISSUE THREAD

**Issue by mattiasb** [transmission-remote-gtk/transmission-remote-gtk#31]
I just learned that transmission-remote-gtk stores password in clear text under .config/. Please consider using the user keychain instead.

**Comment by snaunton** [comment]
Patch which provides support for libsecret (which replaces gnome-keyring). Seems to work fine on debian testing/buster.

[transmission-remote-gtk-libsecret.patch.gz](https://github.com/transmission-remote-gtk/transmission-remote-gtk/files/1866366/transmission-remote-gtk-libsecret.patch.gz)

**Comment by TingPing** [comment]
Dumping a tarball is not how you submit a patch.

Anyway I've actually grown to dislike using libsecret. It is *less* secure if you use sandboxing like Snap/Flatpak/Firejail, it is a total PITA for anybody not on GNOME, and the security it offers is questionable if you don't have disk encryption anyway which makes the plain text password argument a moot point.

**Comment by evert-mouw** [comment]
Agreed with TingPing. I'm using XFCE and using ext4-encryption for my home directory. I also sync my config files across a few machines. I have no need for this "feature".

**Comment by TingPing** [comment]
I'll just amend my earlier comment, libsecret has fixed the sandbox problem, so secrets are now no longer shared.

It is still a frustrating API to use correctly.

**Comment by somasis** [comment]
I just stumbled upon this issue while looking into how to have a non-plain text password in the config file. Maybe something like a `password-cmd` feature would be nice? Then whatever the user's preference of password storage is could be used anyway without implementing support for any one store in particular; libsecret/gnome-keyring/whatever could just be used through libsecret's own `secret-tool`.

**Comment by ktamail666** [comment]
> Agreed with TingPing. I'm using XFCE and using ext4-encryption for my home directory. I also sync my config files across a few machines. I have no need for this "feature".

With all due respect, I think this is a very poor approach. What if I'm giving a presentation and my monitor is projected? Then there's a password leak. What if I'm sharing my desktop via e.g.: Teams or any other platform and I look at that file? Again, there's a leak. By that logic, we might as well store the root password in plain text because the disk is encrypted anyway.

**Comment by trwnh** [comment]
Full disk encryption doesn't help when you commit your config to a dotfiles repo for the whole world to see.

As it stands, I am **very lucky** I caught this issue before committing anything.

Passwords should not be stored in config files in plaintext. If you're going to store passwords in config files anyway, then at least hash them! But preferably, never store passwords in config files.

**Comment by TingPing** [comment]
Since my earlier comments KWallet has implemented the secret service API. So its slightly better covering both KDE and GNOME users. Its still not the best user experience for everyone else.

Maybe somebody could try cleaning up that patch again. Main issue was that it didn't use the async API.

**Comment by mndscp** [comment]
We're approaching the ten-year mark on this issue :)
I really hope somebody can take a look at this again.

Still ridiculous to have a password in plain text in a config file.

**Comment by LaserEyess** [comment]
I agree that this is an issue that needs to be solved. However, as I mentioned before, solving this for some people and not others is also a ridiculous thing to compromise on. I don't know anyone who uses this on Windows, but I do know a few who use it on macOS. So, is libsecret a reliable solution on those platforms? Is it an acceptable cross platform solution? If not, what else is there?

I am genuinely asking for advice on how best to solve this. I do not know the answer.

**Comment by TingPing** [comment]
Windows and macOS have their own keyring solutions.

libsecret just covers a portion of Linux users.

**Comment by trwnh** [comment]
not sure if these are the correct links but:

- Windows: https://learn.microsoft.com/en-us/windows/win32/secauthn/credentials-management
- macOS: https://developer.apple.com/documentation/security/keychain-services
- Linux: https://www.freedesktop.org/wiki/Specifications/secret-storage-spec/

in Arch Linux, libsecret can be used with org.freedesktop.secrets fulfilled by either gnome-keyring, keepassxc, or kwallet as the secret storage backend.

probably the strategy should look something like this pseudocode:

```py
switch CURRENT_PLATFORM:
  case Platform.WINDOWS:
    use_wincred_api()
  case Platform.MACOS:
    use_macos_keychain_api()
  case Platform.LINUX:
    if ORG_FREEDESKTOP_SECRETS_IS_PRESENT:
      use_fdo_secrets()
    else:
      warn_about_insecure_password_storage()
      store_password_in_memory()
```

The consent dialog / warning would be something like this:

```
WARNING
Could not find a secure secret storage

transmission-remote-gtk requires an implementation of
org.freedesktop.secrets in order to securely store
saved passwords for your connection profiles.

Without a secret storage backend such as
GNOME Keyring, KWallet, or KeePassXC,
transmission-remote-gtk will store your password
in memory only, so you will have to re-enter your
password every time transmission-remote-gtk is
started.

[Dismiss warning]
```
